### ISSUE TITLE
State isn't consistently stored when assigned to variables

### ISSUE THREAD

**Issue by raffaelecostanzo** [rrousselGit/riverpod#3258]
**Describe the bug**
State isn't consistently stored when assigned to variables
I have a todo app with a list of tasks.
If you swipe a task, the task gets deleted and a snackbar lets you undo the deletion, if you do it works properly.
If you tap on a task, you are taken to a form where you can edit the title of the task, onced edited you can press a button to save the changes. Once saved, you are taken to the initial screen and the task is successfully updated. A snackbar lets you undo the update, if you do so, the task doesn't restore properly.

I have a Notifier which emits a List of Tasks, and I set an attribute called "previousState" to store the previous state before updating it on every action.

**To Reproduce**
Create and empy Flutter app and call it "todo_test".
Replace the pubspec.yaml with the following code:
```
name: todo_test
description: "A new Flutter project."
publish_to: 'none'
version: 0.1.0

environment:
  sdk: '>=3.2.3 <4.0.0'

dependencies:
  flutter:
    sdk: flutter
  flutter_localizations:
    sdk: flutter
  flutter_riverpod: ^2.4.5
  intl: any
  path_provider: ^2.1.1
  uuid: ^4.2.1

dev_dependencies:
  flutter_test:
    sdk: flutter
  flutter_lints: ^3.0.1

flutter:
  uses-material-design: true

```
Replace the main.dart file with the following code:
```
import 'package:flutter/material.dart';
import 'package:uuid/uuid.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() {
  runApp(
    const ProviderScope(
      child: MainApp(),
    ),
  );
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(
        body: Center(child: TaskList()),
      ),
    );
  }
}

const uuid = Uuid();

// this is the model of the task containing an id and a title
class TaskModel {
  TaskModel({
    required this.title,
  }) : id = uuid.v4();

  TaskModel.forUpdate(String taskId, TaskModel task)
      : id = taskId,
        title = task.title;

  final String id;
  final String title;
}

// this is the task provider. the state emits a list of task models
class TasksNotifier extends Notifier<List<TaskModel>> {
  List<TaskModel> _fetchTasks() {
    return [TaskModel(title: 'Brush your teeth')];
  }

  @override
  List<TaskModel> build() {
    return _fetchTasks();
  }

  late List<TaskModel> previousState;

  // update and delete always store the current state to previousState
  // before actually modifying the state
  // so that I can always call undo() for every action called
  // and restore the previous state
  void updateTask(String taskId, TaskModel updatedTask) {
    debugPrint("updateTask called");
    previousState = state;
    final localPreviousState = state;
    final taskToReplaceIndex = state.indexWhere((task) => task.id == taskId);
    localPreviousState[taskToReplaceIndex] =
        TaskModel.forUpdate(taskId, state[taskToReplaceIndex]);
    state = localPreviousState;
  }

  void deleteTask(String taskId) {
    debugPrint("deleteTask called");
    previousState = state;
    state = state.where((task) => task.id != taskId).toList();
  }

  void undo() {
    debugPrint("undo called");
    final localPreviousState = state;
    for (final e in previousState) {
      debugPrint("undo - previousState - ${e.title}");
    }
    for (final e in state) {
      debugPrint("undo - state - ${e.title}");
    }
    state = previousState;
    previousState = localPreviousState;
  }
}

final tasksProvider = NotifierProvider<TasksNotifier, List<TaskModel>>(
  TasksNotifier.new,
);

// this is the task item used in the task list below
// it is a dismissable so you can swipe the tile and delete the task
class TaskItem extends ConsumerWidget {
  const TaskItem({super.key, required this.task});

  final TaskModel task;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Dismissible(
      key: ValueKey(task.id),
      onDismissed: (_) {
        final taskNotifier = ref.read(tasksProvider.notifier);
        taskNotifier.deleteTask(task.id);
        ScaffoldMessenger.of(context).clearSnackBars();
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: const Text('Task deleted successfully'),
            action: SnackBarAction(
              label: 'Undo',
              onPressed: () => taskNotifier.undo(),
            ),
          ),
        );
      },
      child: ListTile(
        title: Text(task.title),
        onTap: () => Navigator.push(
          context,
          MaterialPageRoute(
            builder: (context) => TaskFormView(task: task),
          ),
        ),
      ),
    );
  }
}

// task list
class TaskList extends ConsumerWidget {
  const TaskList({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final tasks = ref.watch(tasksProvider);
    return tasks.isEmpty
        ? const Center(
            child: Text('There no tasks at the moment.'),
          )
        : ListView.builder(
            itemCount: tasks.length,
            itemBuilder: (context, index) => TaskItem(task: tasks[index]),
          );
  }
}

// this is the form of the task containing the task title and
// a button that calls updateTask, goes to the home screen and shows
// a snack bar message with the option to undo the update
class TaskFormView extends ConsumerStatefulWidget {
  const TaskFormView({super.key, this.task});

  final TaskModel? task;

  @override
  ConsumerState<TaskFormView> createState() => _NewTaskViewState();
}

class _NewTaskViewState extends ConsumerState<TaskFormView> {
  final _formKey = GlobalKey<FormState>();
  final _titleTextController = TextEditingController();

  @override
  void initState() {
    if (widget.task != null) {
      _titleTextController.text = widget.task!.title;
    }
    super.initState();
  }

  @override
  void dispose() {
    _titleTextController.dispose();
    super.dispose();
  }

  // this is where the update and undo get called
  void _submit() {
    if (_formKey.currentState!.validate()) {
      _formKey.currentState!.save();

      final taskNotifier = ref.read(tasksProvider.notifier);

      if (widget.task != null) {
        taskNotifier.updateTask(
          widget.task!.id,
          TaskModel.forUpdate(
            widget.task!.id,
            TaskModel(title: _titleTextController.text),
          ),
        );
      }
      Navigator.of(context).pop();
      ScaffoldMessenger.of(context).clearSnackBars();
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: const Text('Task updated successfully'),
          action: SnackBarAction(
            label: 'Undo',
            onPressed: () => taskNotifier.undo(),
          ),
        ),
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.task!.title),
      ),
      body: SizedBox(
        height: double.infinity,
        child: SingleChildScrollView(
          child: Form(
            key: _formKey,
            child: Padding(
              padding: const EdgeInsets.fromLTRB(16, 48, 16, 16),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  TextFormField(
                    controller: _titleTextController,
                    decoration: const InputDecoration(
                      label: Text('Title'),
                      border: OutlineInputBorder(),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _submit,
        child: const Icon(Icons.edit),
      ),
    );
  }
}
```
Run without debugging on an Android emulator with Android 14.0 and API version 34.
Swipe the task to the right or to the left to delete it. Tap undo on the snack bar message in the bottom. It will restores properly.
Tap on the task, change the title of the task, tap on the edit icon in the bottom right corner. Tap undo on the snack bar message in the bottom. The task won't restore properly.

See the code, the comments in the code and the debug printings to understand more.

**Expected behavior**
When calling undo after the updateTask call, the previousState should be restored properly.

**Comment by raffaelecostanzo** [comment]
I edited the issue by removing some unnecessary code and using Notifier with NotifierProvider because I found out that the problem is not the AsyncNotifier.

The problem is when I assign states to variables like this:
```
final localPreviousState = state;
state = previousState;
```
State updates correctly for deleteTask() and undo() works correctly.
It doesn't for updateTask().

I tried the following and it works:
```
final localPreviousState = [...state];
state = [...previousState];
```
If I change all the assignments like this, the code works properly.

Is this an expected behaviour?
